### ISSUE TITLE
format is not defined & help with various image manipulation requirements ?

### ISSUE THREAD

**Issue by robots4life** [sarpisik/strapi-customize-image-formats#1]
Thank you for your **[blog post and tutorial](https://sarpisik.com/blog/how-to-generate-different-image-formats-with-strapi-plugin-upload)**, this has already helped be a great deal. Thank you! :+1: 

When using your code I think I found a tiny typo.
```javascript
[2020-10-22T13:06:15.282Z] error ReferenceError: format is not defined at Object.uploadFileAndPersist
(/shared/httpd/learn-strapi/strapi-demo/extensions/upload/services/Upload.js:39:7)
```
Change `format[0] !== undefined` to `formats[0] !== undefined` and the error is gone.
https://github.com/sarpisik/strapi-customize-image-formats/blob/73a69ef8f9513e39e417c169b9c123e081d75dce/extensions/upload/services/Upload.js#L33

Trying your code with Strapi 3.2.4 and this error gone I can however not see the Base64 sizes being generated. Where are those or how can I access those please?

Apart from your **awesome work**, would you kindly be able to help me with a few requirements regarding image manipulation please? I am trying to get a few other features for the image manipulation going.. :grin: 

1. How can the original uploaded file, in JPEG or PNG format, also be converted to WebP format but not resized?
2. It would be great if the files could be renamed to `filename`-`width`x`height`.`extension` and have the random string removed if possible. If this is not doable, at least the `filename`-`width`x`height` portion in the filename would be of tremendous help.
3. For encoding to WebP, where can I pass encoding options, i.e. `lossless`, to Sharp?

Pretty new to Strapi (and Sharp) and super happy to have found your tutorial and demo code. :smile: 
Proper handling of client images and catering for responsive sizes and better format is a must in my workflow, being able to get this running with Strapi would be ace. :heart: 

Also, when uploading to the Media Library, does Strapi create image objects in the database with image data? It would be fab if that data could also be made available over an endpoint so I can have JSON data of all the uploaded images.

Bottom line, I am just getting started and would be super happy to hear back from you if you kindly find some time.

**Comment by robots4life** [comment]
> Also, when uploading to the Media Library, does Strapi create image objects in the database with image data? It would be fab if that data could also be made available over an endpoint so I can have JSON data of all the uploaded images.

[https://strapi.io/documentation/v3.x/plugins/upload.html#endpoints](https://strapi.io/documentation/v3.x/plugins/upload.html#endpoints) :heavy_check_mark:

**Comment by sarpisik** [comment]
Hi @robots4life, I am glad the article was useful.

Base64 written to the database directly as a string. The tutorial uses the default database(SQLite). If you installed the [graphql plugin](https://strapi.io/documentation/v3.x/plugins/graphql.html), the following query would return all images with created all formats including the base64:
```javascript
query Uploads {
  files {
    url
    formats
  }
}
```

1. You can create a similar function of "resizeTo" for that:
```javascript
const toFormat = (buffer, format) =>
  sharp(buffer)
    .toFormat(format)
    .toBuffer()
    .catch(() => null);
```

2. You can override the implementation for naming the files from "generateBreakpoint" helper function in "image-manipulation.js" file.

3. [This example might help.](https://sharp.pixelplumbing.com/api-output#webp):+1:.

**Comment by robots4life** [comment]
> Base64 written to the database directly as a string.

Yeah, since I made the Upload Endpoint public I can see the Base64 string, thank you. :heavy_check_mark: 

> 1. You can create a similar function of "resizeTo" for that..

The thing is, I would not know how to plug such a function into the service, I am trying to learn this and see what is connected to where. It is hard to explain, I am trying to visualize how the data flows through the service and would not know where such a `toFormat` function would further go.

> 2. You can override the implementation for naming the files from "generateBreakpoint" helper function in "image-manipulation.js" file.

Ok, I can change the file name of the entry in the API here.
https://github.com/sarpisik/strapi-customize-image-formats/blob/upload/extensions/upload/services/image-manipulation.js#L166
```javascript
name: `${file.name}_${key}_.${format}`,
```
results in this for example. And I just see I need to separate the filename and the extension in my pattern.. like you did. :smile: 

![Screenshot_20201023_160458](https://user-images.githubusercontent.com/34193243/97013630-88fcd900-1549-11eb-9a74-951ecff4de3b.png)

However the actual file names on disk still have the default naming pattern.

![Screenshot_20201023_160708](https://user-images.githubusercontent.com/34193243/97013827-d24d2880-1549-11eb-9f3e-e35400b325d9.png)

How can I change the actual file names on **disk** to a pattern of my liking?
I am after this where the `filename` is without `ext`. :grin: 
```javascript
`filename_breakpoint_hash.ext`
```
If this is also done inside `generateBreakpoint`, how please? :thinking: 

> 3.

Yeah, I am looking at the docs of Sharp, hope I get any further with that. :v:

**Comment by robots4life** [comment]
> 2. You can override the implementation for naming the files from "generateBreakpoint" helper function in "image-manipulation.js" file.

With your help and reading this comment https://github.com/strapi/strapi/issues/1805#issuecomment-416167263 I understood that the name of the file being created on disk comes form the **`file.hash`** property.
Extracting the hash of that and then changing the value to a pattern of my liking did the trick. :heavy_check_mark: 

```javascript
const generateBreakpoint = async (key, { file, toFormat, breakpoint }) => {
  const newBuff = await resizeTo(
    file.buffer,
    {
      width: breakpoint,
      height: breakpoint,
      fit: "inside",
    },
    toFormat
  );

  if (newBuff) {
    const data = await getMetadatas(newBuff);
    const { width, height, size } = data;
    const format = reFormat(data.format);
    const ext = `.${format}`;
    // Change the filename on disk.

    console.log(file);

    // Extract file hash : Nasa_Sp-ace_10_1940044cd2
    // https://stackoverflow.com/a/22082519
    // https://stackoverflow.com/a/38276239
    const regex = /([^_.-]+)$/gm;
    const str = file.hash;
    const pureHash = regex.exec(str)[0];
    console.log(pureHash);

    // Gets just the filename without file extension.
    const getFilename = (name) =>
      name
        .split(".")
        .slice(0, name.split(".").length - 1)
        .join(".");

    // https://github.com/strapi/strapi/issues/1805#issuecomment-416167263
    // "..define the hash that will became the file name.."
    // Change the value of the hash property to change the filename on disk.
    return {
      key,
      file: {
        // name: `${key}_${reName(file.name, ext)}`,
        name: `${getFilename(file.name)}_${key}.${format}`,
        // hash: `${key}_${file.hash}`,
        hash: `${getFilename(file.name)}_${key}_${pureHash}`,
        ext,
        mime: `image/${format}`,
        width,
        height,
        size: bytesToKbytes(size),
        buffer: newBuff,
        path: file.path ? file.path : null,
      },
    };
  }
};
```
![Screenshot_20201024_015633](https://user-images.githubusercontent.com/34193243/97062648-326cba80-159c-11eb-8866-9bd8c8dc7c2d.png)

![Screenshot_20201024_160526](https://user-images.githubusercontent.com/34193243/97083784-cd07e080-1612-11eb-82c3-bfbbe634cc34.png)

With this in place and naming the breakpoints to whatever I want I can get just the filename pattern I need.
:heart: :rocket: :sunglasses:

**Comment by robots4life** [comment]
> 3. For encoding to WebP, where can I pass encoding options, i.e. lossless, to Sharp?

Hi @sarpisik I managed to pass encoding options to `.toFormat` as 2nd argument of [that method of Sharp](https://sharp.pixelplumbing.com/api-output#toformat). :heavy_check_mark: 

```js
// https://sharp.pixelplumbing.com/api-output#toformat
// as 2nd argument we can pass the encoding options for the format :)
const resizeTo = (buffer, options, format) =>
  sharp(buffer)
    .resize(options)
    .toFormat(format, { quality: 80, nearLossless: true, reductionEffort: 6 }) // passing webp and jpg encoding options to .toFormat
    .toBuffer()
    // .catch(() => null);
    // seems to work, no errors
    .catch((error) => console.log(error));
```
Since `format` changes from i.e. `jpg` to `webp` but they have different encoding options I "guess" that if an encoding option is passed that does not exist for the format it is skipped since no errors are logged. Though I am not entirely confident here. Do you know more about this?

And well, now the only thing missing is being able to also encode the original uploaded file, though **without** resizing it. Would you kindly be able to help me with that please?

**Comment by sarpisik** [comment]
@robots4life,

> And well, now the only thing missing is being able to also encode the original uploaded file, though without resizing it. Would you kindly be able to help me with that please?

Following changes would do the job:
```javascript
const generateResponsiveFormats = async (file) => {
  const {
    responsiveDimensions = false,
  } = await strapi.plugins.upload.services.upload.getSettings();

  if (!responsiveDimensions) return [];

  if (!(await canBeProccessed(file.buffer))) return [];

  const originalDimensions = await getDimensions(file.buffer);
  const format = getFormat(file.ext);
  const toFormats = [format];

  // We check the format of the original image to prevent
  // duplication of the same format.
  // Because in default, we create "webp" as an additional format.
  if (format !== "webp") {
    toFormats.push("webp");
  }
  return await Promise.all(
    toFormats.map((toFormat) => {
      // For each format,
      return Promise.all(
        BREAKPOINTS.map(([key, breakpoint]) => {
          if (breakpointSmallerThan(breakpoint, originalDimensions)) {
            // Generate breakpoint sized image
            return generateBreakpoint(key, {
              file,
              toFormat,
              breakpoint,
              originalDimensions,
            });
          }
        }).concat([
          // Compress the original file without rezsizing.
          compressOriginal(file),
        ])
      );
    })
  );
};
```

This is the actual function which compressing without rezising:
```javascript
const compressOriginal = async (originalFile) => {
  const compressedBuff = await sharp(originalFile.buffer)
    .jpeg({ quality: 70, force: false })
    .png({ quality: 70, force: false })
    .webp({ quality: 70, force: false })
    .toBuffer();
  return createFile("original", originalFile, compressedBuff);
};
```

I extracted the creating file metadata part from "generateBreakpoint" function for reusability:
```javascript
const createFile = async (key, file, buffer) => {
  const data = await getMetadatas(buffer);
  const { width, height, size } = data;
  const format = reFormat(data.format);
  const ext = `.${format}`;

  return {
    key,
    file: {
      name: `${key}_${reName(file.name, ext)}`,
      hash: `${key}_${file.hash}`,
      ext,
      mime: `image/${format}`,
      width,
      height,
      size: bytesToKbytes(size),
      buffer,
      path: file.path ? file.path : null,
    },
  };
};
```

and this is the updated version of "generateBreakpoint":
```javascript
const generateBreakpoint = async (key, { file, toFormat, breakpoint }) => {
  const newBuff = await resizeTo(
    file.buffer,
    {
      width: breakpoint,
      height: breakpoint,
      fit: "inside",
    },
    toFormat
  );

  if (newBuff) return createFile(key, file, newBuff);
};
```

**Comment by robots4life** [comment]
@sarpisik Thank you very much. I will still get back to you about the code, quite busy, so have not had the time to check this out. However it is very nice of you to help me with this. :+1:

**Comment by kwiat1990** [comment]
Hi,
I have also copied your methods into my Strapi project and I have already change the first problematic part `format[0] !== undefined` to `formats[0] !== undefined` but then I saw just another error: `TypeError: Cannot destructure property 'key' of '.for' as it is undefined.` in `Upload.js`. 

Right now I can't upload any image and I have absolutely no idea what's wrong with the code. Simultaneously I can see that inside `public/upload` all images (besides base64 variants) are created.

I can see in the console, that `formats` Array contains only undefined items `format [ undefined, undefined, undefined ]`.

Have you guys any advice how I could fix it?

Another question is: do I need all methods (I mean to create large, medium and small variant) or as far as it is automatically handled by Strapi (v3.3.4) I need only to implement one for the base64 variant?

**Comment by robots4life** [comment]
@kwiat1990 I still have to go over the code but as the holiday is approaching I will have time and look more than forward to give all this a good testing. If you can wait a tiny bit longer I will get back to you with working code.

The features I am looking for are the following.
1. rename the created images to a pattern I like, I will explain to you how this can be changed
2. define the different formats and sizes of the created image, so jpeg, png or webp and the dimensions of the image
3. Understand how this computation work could be outsourced to Digital Ocean for example.
https://strapi.io/documentation/3.0.0-beta.x/plugins/upload.html
https://medium.com/swlh/resize-and-upload-your-images-to-digital-ocean-spaces-with-strapi-js-edf0af1b4d88
Once I figure this out and have a working example using a local provider as well as an external provider I will get back to you.

**Comment by ahneemah** [comment]
Any updates on this?
